File: bulletin_board/thresholdalgs.py
import math
import hashlib
import logging

from helios.crypto import randpool, number

from helios.crypto import numtheory
from helios.crypto import elgamal
import sys
from helios.crypto.algs import Utils
from helios.crypto import algs
from helios.crypto import utils

from fractions import *


class Thresholdscheme():

    def __init__(self, n=None, k=None, ground_1=None, ground_2=None):
        self.ground_1 = ground_1
        self.ground_2 = ground_2
        self.n = n
        self.k = k

    def share_verifiably(self, s, t, EG, trustees):
        q = EG.q
        p = EG.p
        g = EG.g
        F = Polynomial(s, self, EG)
        G = Polynomial(t, self, EG)

        # Create points on polynomials from trustee x values.
        # F(0) is secret!
        points_F = F.create_points(trustees)
        points_G = G.create_points(trustees)

        # create commitments
        Ei = []
        for i in range(self.k):
            commitment_loop = Commitment_E()
            commitment_loop.generate(F.coeff[i], G.coeff[i], self.ground_1, self.ground_2, p, q, g)
            if commitment_loop.value > p - 1:
                raise Exception('Ei value to big!')

            Ei.append(commitment_loop)

        shares = []
        for i in range(self.n):
            share = Share(points_F[i], points_G[i], Ei)
            if share.verify_share(self, p, q, g):
                shares.append(share)
            else:
                return None

        return shares


class Polynomial():
    # c0 is the free term and is equal to coeff[0]
    # grade is k-1

    def __init__(self, c0=None, scheme=None, EG=None):
        p = EG.p
        q = EG.q

        self.coeff = []
        self.coeff.append(c0 % q)
        self.EG = EG
        self.scheme = scheme
        self.grade = self.scheme.k - 1

        for i in range(self.grade):
            # coefficients should be mod q
            self.coeff.append(Utils.random_mpz_lt(q))

    def set(self, coeff):
        self.coeff = coeff

    def evaluate(self, x):
        q = self.EG.q
        p = self.EG.p
        value = 0
        for i in range(self.grade + 1):
            value = (value + (self.coeff[i] * pow(x, i, q))) % q
        return value

    def create_points(self, trustees):
        points = []
        for trustee in trustees:
            point = Point(trustee.id, self.evaluate(trustee.id))
            points.append(point)

        return points


class Point():

    def __init__(self, x, y):
        self.x_value = x
        self.y_value = y

    def encrypt(self, public_key):
        plain_x = algs.EGPlaintext(self.x_value, public_key)
        plain_y = algs.EGPlaintext(self.y_value, public_key)

        return Encrypted_point(public_key.encrypt(plain_x), public_key.encrypt(plain_y))

    def on_polynomial(self, polynom):
        if polynom.evaluate(self.x_value) == self.y_value:
            return True
        else:
            return False

    @classmethod
    def from_dict(cls, d):
        """
        Deserialize from dictionary.
        """
        x_value = int(d['x_value'])
        y_value = int(d['y_value'])
        point = cls(x_value, y_value)

        return point

    def to_dict(self):
        """
        Serialize to dictionary.
        """
        return {'x_value': str(self.x_value), 'y_value': str(self.y_value)}


class Encrypted_point():

    def __init__(self, ciph_x=None, ciph_y=None):
        self.ciph_x = ciph_x
        self.ciph_y = ciph_y

    def decrypt(self, secret_key):
        x_val = secret_key.decrypt(self.ciph_x).m
        y_val = secret_key.decrypt(self.ciph_y).m

        return Point(x_val, y_val)

    @classmethod
    def from_dict(cls, d):
        """
        Deserialize from dictionary.
        """
        ciph_x = algs.EGCiphertext.from_dict(d['ciph_x'])
        ciph_y = algs.EGCiphertext.from_dict(d['ciph_y'])
        encr_point = Encrypted_point(ciph_x, ciph_y)
        return encr_point

    def to_dict(self):
        """
        Serialize to dictionary.
        """
        return {'ciph_x': self.ciph_x.to_dict(), 'ciph_y': self.ciph_y.to_dict()}


class Share():

    def __init__(self, point_s=None, point_t=None, Ei=None):
        self.point_s = point_s
        self.point_t = point_t
        self.Ei = Ei

    def add(self, addedshare, p, q, g):
        if self.point_s.x_value == addedshare.point_s.x_value == self.point_t.x_value == addedshare.point_t.x_value:
            x = self.point_s.x_value
            new_point_s = Point(x, (self.point_s.y_value + addedshare.point_s.y_value) % q)
            new_point_t = Point(x, (self.point_t.y_value + addedshare.point_t.y_value) % q)
            new_Ei = []
            for i in range(len(self.Ei)):
                Ei_now = self.Ei[i]
                Ei_now.add(addedshare.Ei[i], p, q, g)
                new_Ei.append(Ei_now)
            self.point_s = new_point_s
            self.point_t = new_point_t
            self.Ei = new_Ei

    def encrypt(self, public_key):
        encry_point_s = self.point_s.encrypt(public_key)
        encry_point_t = self.point_t.encrypt(public_key)
        encry_share = Encrypted_Share(encry_point_s, encry_point_t, self.Ei)

        return encry_share

    def sign(self, secret_key, p, q, g):
        string = utils.to_json_js(self.to_dict())
        sig = Signature()
        sig.generate(string, secret_key, p, q, g)
        return sig

    def verify_share(self, scheme, p, q, g):
        k = scheme.k
        n = scheme.n
        E = Commitment_E()
        E.generate(self.point_s.y_value, self.point_t.y_value, scheme.ground_1, scheme.ground_2, p, q, g)
        if E.value > p - 1:
            raise Exception('E value too big')

        point_number = self.point_s.x_value
        result = 1
        if len(self.Ei) != k:
            raise Exception('Ei has a wrong number of elements')

        for j in range(len(self.Ei)):
            i = point_number
            interm = pow(i, j, p)
            result = (result * pow(self.Ei[j].value, interm, p)) % p

        if result != E.value:
            return False
        else:
            return True

    @classmethod
    def from_dict(cls, d):
        """
        Deserialize from dictionary.
        """
        point_s = Point.from_dict(d['point_s'])
        point_t = Point.from_dict(d['point_t'])
        Ei_dict = d['Ei']
        Ei = []
        i = 0
        while Ei_dict.has_key(str(i)):
            dict = Ei_dict[str(i)]
            com = Commitment_E().from_dict(dict)
            Ei.append(com)
            i = i + 1

        share = Share(point_s, point_t, Ei)
        return share

    def to_dict(self):
        """
        Serialize to dictionary.
        """
        com_dict = {}
        for i in range(len(self.Ei)):
            com_dict[str(i)] = self.Ei[i].to_dict()

        return {'point_s': self.point_s.to_dict(), 'point_t': self.point_t.to_dict(), 'Ei': com_dict}


class Encrypted_Share():

    def __init__(self, encry_point_s=None, encry_point_t=None, Ei=None):
        self.encry_point_s = encry_point_s
        self.encry_point_t = encry_point_t

        self.Ei = Ei

    def decrypt(self, secret_key):
        point_s = self.encry_point_s.decrypt(
            secret_key)  # select only the value
        point_t = self.encry_point_t.decrypt(secret_key)

        return Share(point_s, point_t, self.Ei)

    @classmethod
    def from_dict(cls, d):
        """
        Deserialize from dictionary.
        """
        encry_point_s = Encrypted_point.from_dict(d['encry_point_s'])
        encry_point_t = Encrypted_point.from_dict(d['encry_point_t'])
        Ei_dict = d['Ei']
        Ei = []
        i = 0
        while Ei_dict.has_key(str(i)):
            dict = Ei_dict[str(i)]
            com = Commitment_E().from_dict(dict)
            Ei.append(com)
            i = i + 1

        encry_share = Encrypted_Share(encry_point_s, encry_point_t, Ei)

        return encry_share

    def to_dict(self):
        """
        Serialize to dictionary.
        """
        com_dict = {}
        for i in range(len(self.Ei)):
            com_dict[str(i)] = self.Ei[i].to_dict()

        return {'encry_point_s': self.encry_point_s.to_dict(), 'encry_point_t': self.encry_point_t.to_dict(), 'Ei': com_dict}


class Signed_Encrypted_Share():

    def __init__(self, sig=None, encr_share=None):
        self.sig = sig
        self.encr_share = encr_share

    @classmethod
    def from_dict(cls, d):
        """
        Deserialize from dictionary.
        """
        encr_share = Encrypted_Share.from_dict(d['encr_share'])
        sig = Signature.from_dict(d['sig'])
        sig_encr_share = Signed_Encrypted_Share(sig, encr_share)

        return sig_encr_share

    def to_dict(self):
        """
        Serialize to dictionary.
        """
        return {'sig': self.sig.to_dict(), 'encr_share': self.encr_share.to_dict()}


class Committed_Point():

    def __init__(self, point, E, Ei):
        self.point = point
        self.E = E
        self.Ei = Ei  # list of commitments

    def verify_commitments(self):
        p = self.E.p
        point_number = self.point.x_value
        result = 1
        for j in range(len(self.Ei)):
            i = point_number
            interm = pow(i, j, p)
            result = (result * pow(self.Ei[j].value, interm, p)) % (p - 1)
            if result != self.E.value:
                return False
            else:
                return True


class Feedback():

    def __init__(self, trustee_id, feedback):
        self.trustee_id = trustee_id
        self.feedback = feedback

    def set_feedback(self, feedback):
        self.feedback = feedback


class Commitment_E():

    def __init__(self, ground_1=None, ground_2=None, value=None):
        self.ground_1 = ground_1
        self.ground_2 = ground_2
        self.value = value

    def generate(self, s, t, ground_1, ground_2, p, q, g):
        s = s % q
        t = t % q
        self.ground_1 = ground_1
        self.ground_2 = ground_2
        self.value = (pow(self.ground_1, s, p) * pow(self.ground_2, t, p)) % p

    def add(self, addedcommitment, p, q, g):
        if self.ground_1 == addedcommitment.ground_1 and self.ground_2 == addedcommitment.ground_2:
            self.value = (self.value * addedcommitment.value) % p

    @classmethod
    def from_dict(cls, d):
        """
        Deserialize from dictionary.
        """
        ground_1 = int(d['ground_1'])
        ground_2 = int(d['ground_2'])
        value = int(d['value'])
        com = Commitment_E(ground_1, ground_2, value)

        return com

    @classmethod
    def array_from_dict(cls, d):
        i = 0
        while d.has_key(str(i)):
            dict = d[str(i)]
            com = Commitment_E().from_dict(dict)
            com = Commitment_E()
            i = i + 1

    def to_dict(self):
        """
        Serialize to dictionary.
        """
        return {'ground_1': str(self.ground_1), 'ground_2': str(self.ground_2), 'value': str(self.value)}


class Trustee():

    def __init__(self, trustee_id, public_key=None):
        self.trustee_id = trustee_id
        self.public_key = public_key

    def generate_keypair(self, ELGAMAL_PARAMS):
        keypair = ELGAMAL_PARAMS.generate_keypair()
        self.public_key = keypair.pk
        self.secret_key = keypair.sk
        self.pok = self.secret_key.prove_sk(algs.DLog_challenge_generator)

    def write_public_key(self, path):
        y = self.public_key.y
        string = utils.to_json(self.public_key.to_dict())
        f = open(path, 'w')
        f.write(string + '\n')
        f.close

    def write_secret_key(self, path):
        x = self.secret_key.x
        #string = str(x)+'\n'
        string = utils.to_json(self.secret_key.to_dict())
        f = open(path, 'w')
        f.write(string + '\n')
        f.close()

    def write_pok(self, path):
        pok = self.pok
        string = utils.to_json(self.pok.to_dict())
        f = open(path, 'w')
        f.write(string + '\n')
        f.close()


class Signature():

    def __init__(self, r=None, s=None):
        self.r = r
        self.s = s

    def generate(self, m, secret_key, p, q, g):
        while True:
            hash = utils.hash_b64(m)
            hash_dec = utils.encode_string_to_decimal(hash)
            k = algs.Utils.random_k_relative_prime_p_1(p)
            k_inv = algs.Utils.inverse(k, p - 1)
            x = secret_key.x
            r = pow(g, k, p)
            s = ((hash_dec - x * r) * k_inv) % (p - 1)
            if s != 1:
                self.s = s
                self.r = r
                return None

    def verify(self, m, public_key, p, q, g):
        correct = True
        y = public_key.y
        hash = utils.hash_b64(m)
        hash_dec = utils.encode_string_to_decimal(hash)
        if self.r >= p:
            correct = False
        if self.s >= p - 1:
            correct = False
        val = (pow(y, self.r, p) * pow(self.r, self.s, p)) % p
        if (pow(g, hash_dec, p) != val):
            correct = False

        return correct

    @classmethod
    def from_dict(cls, d):
        """
        Deserialize from dictionary.
        """
        r = int(d['r'])
        s = int(d['s'])

        return cls(r, s)

    def to_dict(self):
        """
        Serialize to dictionary.
        """
        return {'r': str(self.r), 's': str(self.s)}

File: helios/media/helios/jscrypto/thresholdalgs.js
thresholdalgs = {};

thresholdalgs.encode_string_to_decimal = function (s) {
    var tmp = '';
    for (var i = 0; i < s.length; i++) {
        var c = s[i];
        tmp = tmp + String(c.charCodeAt(0));
    }
    return BigInt.fromJSONObject(tmp)
}

thresholdalgs.Polynomial = Class.extend({
    init: function (c0, scheme, EG) {
        var p = EG.p;
        var q = EG.q;
        this.coeff = [];
        this.coeff.push(c0);
        this.EG = EG;
        this.scheme = scheme;
        this.grade = this.scheme.k - 1;

        for (var i = 0; i < this.grade; i++) {
            this.coeff.push(Random.getRandomInteger(q));
        }
    },

    set: function (coeff) {
        this.coeff = coeff;
        return null;
    },

    create_points: function (ids) {
        n = this.scheme.n;
        var points = [];
        for (var i = 0; i < n; i++) {
            var point = new thresholdalgs.Point(BigInt.fromInt(ids[i]), this.evaluate(ids[i]));
            points.push(point);
        }
        return points;

    },
    evaluate: function (x) {
        var q = this.EG.q;
        var p = this.EG.p;
        var x = BigInt.fromInt(x);
        var value = BigInt.ZERO;
        for (i = 0; i < this.grade + 1; i++) {
            var pow = BigInt.fromInt(i);
            value = (value.add(this.coeff[i].multiply(x.modPow(pow, q))));
            value = value.mod(q);
        }
        return value;

    }
});

thresholdalgs.Point = Class.extend({
    init: function (x, y) {
        this.x_value = x;
        this.y_value = y;
    },

    encrypt: function (public_key) {
        var plain_x = new ElGamal.Plaintext(this.x_value, public_key);
        var plain_y = new ElGamal.Plaintext(this.y_value, public_key);
        var ciph_x = ElGamal.encrypt(public_key, plain_x);
        var ciph_y = ElGamal.encrypt(public_key, plain_y);
        return new thresholdalgs.EncryptedPoint(ciph_x, ciph_y);
    },

    on_polynomial: function (polynom) {
        if (polynom.evaluate(this.x_value) == this.y_value)
            return true;
        else
            return false;
    },

    toJSONObject: function () {
        return {
            x_value: this.x_value.toJSONObject(),
            y_value: this.y_value.toJSONObject()
        };
    },

});

thresholdalgs.Point.fromJSONObject = function (d) {
    return new thresholdalgs.Point(d.x_value, d.y_value);
};

thresholdalgs.EncryptedPoint = Class.extend({
    init: function (ciph_x, ciph_y) {
        this.ciph_x = ciph_x;
        this.ciph_y = ciph_y;
    },

    decrypt: function (secret_key) {
        this.ciph_x.pk = secret_key.pk;
        this.ciph_y.pk = secret_key.pk;
        var x_val = secret_key.decrypt(this.ciph_x).m;
        var y_val = secret_key.decrypt(this.ciph_y).m;
        return new thresholdalgs.Point(x_val, y_val);
    },

    toJSONObject: function () {
        return {
            ciph_x: this.ciph_x.toJSONObject(),
            ciph_y: this.ciph_y.toJSONObject()
        };

    }
});

thresholdalgs.EncryptedPoint.fromJSONObject = function (d) {
    return new thresholdalgs.EncryptedPoint(ElGamal.Ciphertext.fromJSONObject(d.ciph_x), ElGamal.Ciphertext.fromJSONObject(d.ciph_y));
};

thresholdalgs.Share = Class.extend({
    init: function (point_s, point_t, Ei) {
        this.point_s = point_s;
        this.point_t = point_t;
        this.Ei = Ei;
    },

    add: function (addedshare, p, q, g, scheme) {
        if (this.point_s.x_value.equals(addedshare.point_s.x_value) && this.point_s.x_value.equals(addedshare.point_t.x_value) && this.point_s.x_value.equals(this.point_t.x_value)) {
            x = this.point_s.x_value;
            var new_point_s = new thresholdalgs.Point(x, (this.point_s.y_value.add(addedshare.point_s.y_value)).mod(q));
            var new_point_t = new thresholdalgs.Point(x, (this.point_t.y_value.add(addedshare.point_t.y_value)).mod(q));
            new_Ei = [];
            var overall_succes = true;
            for (var i = 0; i < this.Ei.length; i++) {
                Ei_now = this.Ei[i];
                var succes = Ei_now.add(addedshare.Ei[i], p, q, g, scheme);
                if (!succes)
                    overall_succes = false;

                new_Ei.push(Ei_now);
            }
            this.point_s = new_point_s;
            this.point_t = new_point_t;
            this.Ei = new_Ei;

            if (this.verify_share(scheme, p, q, g)) {
                if (overall_succes) {
                    return true;
                } else {
                    return false;
                }
            } else {
                return false;
            }
        } else {
            return false;
        }
    },

    encrypt: function (public_key) {
        var encry_point_s = this.point_s.encrypt(public_key);
        var encry_point_t = this.point_t.encrypt(public_key);
        var encry_share = new thresholdalgs.EncryptedShare(encry_point_s, encry_point_t, this.Ei);
        return encry_share;
    },

    get_json_string: function () {
        return JSON.stringify(this.toJSONObject());

    },
    sign: function (secret_key, p, q, g) {
        var sig = new thresholdalgs.Signature(null, null);
        sig.generate(this.get_json_string(), secret_key, p, q, g);
        return sig;
    },

    verify_share: function (scheme, p, q, g) {
        var k = scheme.k;
        var n = scheme.n;
        var E = new thresholdalgs.CommitmentE(null, null, null)
        E.generate(this.point_s.y_value, this.point_t.y_value, scheme.ground_1, scheme.ground_2, p, q, g);
        var point_number = this.point_s.x_value;
        var result = BigInt.ONE;
        if (this.Ei.length != k)
            return false;
        for (var j = 0; j < this.Ei.length; j++) {
            var i = point_number;
            var pow = BigInt.fromInt(j);
            var interm = i.modPow(pow, p).mod(p);
            result = (result.multiply(BigInt.fromInt(this.Ei[j].value).modPow(interm, p))).mod(p);
        }
        if (!result.equals(E.value)) {
            return false;
        } else {
            return true;
        }
    },

    toJSONObject: function () {
        var com_dict = {};
        for (var i = 0; i < this.Ei.length; i++) {
            com_dict[String(i)] = this.Ei[i].toJSONObject();
        }
        return {
            Ei: com_dict,
            point_s: this.point_s.toJSONObject(),
            point_t: this.point_t.toJSONObject()
        };
    }
});

thresholdalgs.Share.fromJSONObject = function (d) {
    var point_s = Point.fromJSONObject(d.point_s);
    var point_t = Point.fromJSONObject(d.point_t);
    var Ei_dict = d.Ei;
    var Ei = [];
    var i = 0;
    while (String(i) in Ei_dict) {
        var dict = Ei_dict[String(i)];
        var com = new thresholdalgs.CommitmentE.fromJSONObject(dict);
        Ei.push(com);
        i++;
    }
    var share = new thresholdalgs.Share(point_s, point_t, Ei);
    return share;
};

thresholdalgs.EncryptedShare = Class.extend({
    init: function (encry_point_s, encry_point_t, Ei) {
        this.encry_point_s = encry_point_s;
        this.encry_point_t = encry_point_t;
        this.Ei = Ei;
    },

    decrypt: function (secret_key) {
        var point_s = this.encry_point_s.decrypt(secret_key);
        var point_t = this.encry_point_t.decrypt(secret_key);

        return new thresholdalgs.Share(point_s, point_t, this.Ei);
    },

    toJSONObject: function () {
        var com_dict = {};
        for (var i = 0; i < this.Ei.length; i++) {
            com_dict[String(i)] = this.Ei[i].toJSONObject();
        }

        return {
            'Ei': com_dict,
            'encry_point_s': this.encry_point_s.toJSONObject(),
            'encry_point_t': this.encry_point_t.toJSONObject()
        };
    }
});

thresholdalgs.EncryptedShare.fromJSONObject = function (d) {
    var encry_point_s = thresholdalgs.EncryptedPoint.fromJSONObject(d['encry_point_s']);
    var encry_point_t = thresholdalgs.EncryptedPoint.fromJSONObject(d['encry_point_t']);
    var Ei_dict = d['Ei'];
    var Ei = [];
    var i = 0;
    while (String(i) in Ei_dict) {
        var dict = Ei_dict[String(i)];
        var com = new thresholdalgs.CommitmentE.fromJSONObject(dict);
        Ei.push(com);
        i++;
    }
    var encry_share = new thresholdalgs.EncryptedShare(encry_point_s, encry_point_t, Ei);
    return encry_share;
};

thresholdalgs.SignedEncryptedShare = Class.extend({
    init: function (sig, encr_share) {
        this.sig = sig;
        this.encr_share = encr_share;

    },

    toJSONObject: function () {

        return {
            'encr_share': this.encr_share.toJSONObject(),
            'sig': this.sig.toJSONObject()
        };
    },
});

thresholdalgs.SignedEncryptedShare.fromJSONObject = function (d) {
    var encr_share_dict = d['encr_share']
    var encr_share = thresholdalgs.EncryptedShare.fromJSONObject(encr_share_dict);
    var sig = thresholdalgs.Signature.fromJSONObject(d['sig']);
    var sig_encr_share = new thresholdalgs.SignedEncryptedShare(sig, encr_share);

    return sig_encr_share;
};

thresholdalgs.CommitmentE = Class.extend({
    init: function (ground_1, ground_2, value) {
        this.ground_1 = ground_1;
        this.ground_2 = ground_2;
        this.value = value;
    },

    generate: function (s, t, ground_1, ground_2, p, q, g) {
        var s = s.mod(q)
        var t = t.mod(q)

        this.ground_1 = ground_1;
        this.ground_2 = ground_2;
        this.value = ((this.ground_1.modPow(s, p)).multiply(this.ground_2.modPow(t, p))).mod(p);

        return null;
    },

    add: function (addedcommitment, p, q, g, scheme) {
        if ((this.ground_1.equals(addedcommitment.ground_1)) && (this.ground_2.equals(addedcommitment.ground_2))) {
            this.value = (this.value.multiply(addedcommitment.value)).mod(p);
            return true;
        } else {
            return false;
        }
    },
    toJSONObject: function () {
        return {
            'ground_1': this.ground_1.toJSONObject(),
            'ground_2': this.ground_2.toJSONObject(),
            'value': this.value.toJSONObject()
        };

    },
});

thresholdalgs.CommitmentE.fromJSONObject = function (d) {
    var ground_1 = BigInteger.fromJSONObject(d['ground_1']);
    var ground_2 = BigInteger.fromJSONObject(d['ground_2']);
    var value = BigInteger.fromJSONObject(d['value']);

    var com = new thresholdalgs.CommitmentE(ground_1, ground_2, value);

    return com;
};

thresholdalgs.Signature = Class.extend({
    init: function (r, s) {
        this.r = r;
        this.s = s;
    },

    generate: function (m, secret_key, p, q, g) {
        while (true) {
            var p_1 = p.add(BigInt.ONE.negate());
            var hash = b64_sha256(m);

            var hash_dec = thresholdalgs.encode_string_to_decimal(hash);

            var k = Random.random_k_relative_prime_p_1(p);
            var k_inv = k.modInverse(p_1);
            var x = secret_key.x;
            var r = g.modPow(k, p).mod(p);
            var s = (hash_dec.add(x.multiply(r).negate()).multiply(k_inv)).mod(p_1);
            if (!s.equals(BigInt.ONE)) {
                this.s = s;
                this.r = r;
                return null;
            }
        }
    },

    verify: function (m, public_key, p, q, g) {
        var correct = true;
        var y = public_key.y;
        var hash = b64_sha256(m);
        var hash_dec = thresholdalgs.encode_string_to_decimal(hash);

        var val = (y.modPow(this.r, p).multiply(this.r.modPow(this.s, p))).mod(p);
        if (!g.modPow(hash_dec, p).mod(p).equals(val))
            correct = false;

        return correct;
    },

    toJSONObject: function () {
        return {
            'r': this.r.toJSONObject(),
            's': this.s.toJSONObject()
        };
    },
});

thresholdalgs.Signature.fromJSONObject = function (d) {
    var r = BigInt.fromJSONObject(d['r']);
    var s = BigInt.fromJSONObject(d['s']);
    return new thresholdalgs.Signature(r, s);
};

thresholdalgs.ThresholdScheme = Class.extend({
    init: function (election_id, n, k, ground_1, ground_2) {
        this.election_id = election_id;
        this.n = n;
        this.k = k;
        this.ground_1 = ground_1;
        this.ground_2 = ground_2;
    },

    share_verifiably: function (s, t, EG, ids) {
        var F = new thresholdalgs.Polynomial(s, this, EG);
        var G = new thresholdalgs.Polynomial(t, this, EG);

        var points_F = F.create_points(ids);
        var points_G = G.create_points(ids);

        var Ei = [];
        for (var i = 0; i < this.k; i++) {
            var commitment_loop = new thresholdalgs.CommitmentE(null, null, null);
            commitment_loop.generate(F.coeff[i], G.coeff[i], this.ground_1, this.ground_2, EG.p, EG.q, EG.g);
            Ei.push(commitment_loop);
        }

        var shares = [];
        for (var i = 1; i < this.n + 1; i++) {
            var share = new thresholdalgs.Share(points_F[i - 1], points_G[i - 1], Ei);
            if (share.verify_share(this, EG.p, EG.q, EG.g)) {
                shares.push(share);
            } else {
                return null;
            }
        }

        return shares;
    }
});

thresholdalgs.ThresholdScheme.fromJSONObject = function (d) {
    var election_id = parseInt(d.election_id);
    var n = parseInt(d.n);
    var k = parseInt(d.k);
    var ground_1 = BigInt.fromJSONObject(d.ground_1);
    var ground_2 = BigInt.fromJSONObject(d.ground_2);

    return new thresholdalgs.ThresholdScheme(election_id, n, k, ground_1, ground_2);
};
